Userland/Libraries/LibGfx/QOILoader.cpp

```diff
@@ -34,50 +34,42 @@ static ErrorOr<QOIHeader> decode_qoi_header(InputMemoryStream& stream)
     return header;
 }
 
-static ErrorOr<Color> decode_qoi_op_rgb(InputMemoryStream& stream, Color pixel)
+static ErrorOr<Color> decode_qoi_op_rgb(InputMemoryStream& stream, u8 tag, Color pixel)
 {
-    u8 bytes[4];
+    VERIFY(tag == QOI_OP_RGB);
+    u8 bytes[3];
     stream >> Bytes { &bytes, array_size(bytes) };
     if (stream.handle_any_error())
         return Error::from_string_literal("Invalid QOI image: end of stream while reading QOI_OP_RGB chunk");
-    VERIFY(bytes[0] == QOI_OP_RGB);
 
     // The alpha value remains unchanged from the previous pixel.
-    return Color { bytes[1], bytes[2], bytes[3], pixel.alpha() };
+    return Color { bytes[0], bytes[1], bytes[2], pixel.alpha() };
 }
 
-static ErrorOr<Color> decode_qoi_op_rgba(InputMemoryStream& stream)
+static ErrorOr<Color> decode_qoi_op_rgba(InputMemoryStream& stream, u8 tag)
 {
-    u8 bytes[5];
+    VERIFY(tag == QOI_OP_RGBA);
+    u8 bytes[4];
     stream >> Bytes { &bytes, array_size(bytes) };
     if (stream.handle_any_error())
         return Error::from_string_literal("Invalid QOI image: end of stream while reading QOI_OP_RGBA chunk");
-    VERIFY(bytes[0] == QOI_OP_RGBA);
-    return Color { bytes[1], bytes[2], bytes[3], bytes[4] };
+    return Color { bytes[0], bytes[1], bytes[2], bytes[3] };
 }
 
-static ErrorOr<u8> decode_qoi_op_index(InputMemoryStream& stream)
+static ErrorOr<u8> decode_qoi_op_index(InputMemoryStream&, u8 tag)
 {
-    u8 byte;
-    stream >> byte;
-    if (stream.handle_any_error())
-        return Error::from_string_literal("Invalid QOI image: end of stream while reading QOI_OP_INDEX chunk");
-    VERIFY((byte & QOI_MASK_2) == QOI_OP_INDEX);
-    u8 index = byte & ~QOI_MASK_2;
+    VERIFY((tag & QOI_MASK_2) == QOI_OP_INDEX);
```

Comment:
Not entirely happy with the naming here - every byte masked with `QOI_MASK_2` is not actually a 'tag', only two bits are the tag, the rest is payload. Call it `byte` / `first_byte` or something?
(I know this was already the case with `u8 tag = stream.peek_or_error();`, but at least that was actually limited to tag checks.)

Reply:
Fixed!